csharp/src/Microsoft.ML.OnnxRuntime/NamedOnnxValue.cs

```diff
@@ -218,12 +218,22 @@ out nativeTensor
                                                     ));
 
                     // fill the native tensor, using GetValue(index) from the Tensor<string>
-                    string[] stringsInTensor = new string[tensorValue.Length];
+                    var len = tensorValue.Length;
+                    var stringsInTensor = new IntPtr[len];
+                    var pinnedHandles = new GCHandle[len];
                     for (int i = 0; i < tensorValue.Length; i++)
                     {
-                        stringsInTensor[i] = tensorValue.GetValue(i);
+                        var utf8str = UTF8Encoding.UTF8.GetBytes(tensorValue.GetValue(i) + "\0");
+                        pinnedHandles[i] = GCHandle.Alloc(utf8str, GCHandleType.Pinned);
+                        stringsInTensor[i] = pinnedHandles[i].AddrOfPinnedObject();
                     }
-                    NativeApiStatus.VerifySuccess(NativeMethods.OrtFillStringTensor(nativeTensor, stringsInTensor, (UIntPtr)tensorValue.Length));
+
+                    NativeApiStatus.VerifySuccess(NativeMethods.OrtFillStringTensor(nativeTensor, stringsInTensor, (UIntPtr)len));
+                    foreach (var handle in pinnedHandles)
+                    {
+                        handle.Free();
```

Comment:
> handle.Free(); [](start = 23, length = 15)

Can this be handled in an exception safe manner via IDisposable?

Reply:
IDisposable is typically useful for managed objects that hold references to native resources (and hence can free the native resources in the Dispose() call). When ORT returns output tensor values, IDisposable objects are in fact used for this same reason.

However, we don't need to use the IDisposable interface for model inputs, since they are already in managed memory, and don't require any special handling (e.g. freeing native memory, closing out open files etc). 

In this block, we need to pin the arrays before the call to native ORT, and unpin right afterwards (GC will handle the objects correctly afterwards).

Reply:
Let's talk offline and see what happens when pinned memory is not released on exception. Pinner memory means it is subject to memory management and will stay this way and leak.

---
In reply to: [359646557](https://github.com/microsoft/onnxruntime/pull/2697#discussion_r359646557) [](ancestors = 359646557)